aplt/vapid.py

```diff
@@ -0,0 +1,120 @@
+import base64
+import time
+import hashlib
+
+import ecdsa
+import logging
+from jose import jws
+
+
+class VapidException(Exception):
+    pass
+
+
+class Vapid(object):
+    """Minimal VAPID signature generation library. """
+    _private_key = None
+    _public_key = None
+    _hasher = hashlib.sha256
+
+    def __init__(self, private_key_file=None, private_key=None):
+        """Initialize VAPID using an optional file containing a private key
+        in PEM format.
+
+        :param private_key_file: The name of the file containing the
+        private key
+        """
+        if private_key_file:
+            private_key = open(private_key_file).read()
+        if private_key:
+            try:
+                if "BEGIN EC" in private_key:
+                    self._private_key = ecdsa.SigningKey.from_pem(private_key)
+                else:
+                    self._private_key = \
+                        ecdsa.SigningKey.from_der(
+                            base64.standard_b64decode(private_key))
+            except Exception, exc:
+                logging.error("Could not open private key file: %s", repr(exc))
+                raise VapidException(exc)
+            self._pubilcKey = self._private_key.get_verifying_key()
+
+    @property
+    def private_key(self):
+        if not self._private_key:
+            raise VapidException(
+                "No private key defined. Please import or generate a key.")
+        return self._private_key
+
+    @private_key.setter
+    def private_key(self, value):
+        self._private_key = value
+
+    @property
+    def public_key(self):
+        if not self._public_key:
+            self._public_key = self.private_key.get_verifying_key()
+        return self._public_key
+
+    def generate_keys(self):
+        """Generate a valid ECDSA Key Pair."""
+        self.private_key = ecdsa.SigningKey.generate(curve=ecdsa.NIST256p)
+        self.public_key
```

Comment:
self.public_key is a setter. calling this initializes the underlying self._public_key. This bottlenecks to that method instead of duplicating the internals. Is there a more pythonic way to call this?